File: 2_Programação_sistemas_operacionais/2.07_Pipes_Sinais_Alarmes/Ex2.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>

int main()
{
	int pid;	// PID do processo filho
	int fd[2];	// Descritores do pipe
	char mensagem[30];
	pipe(fd); // Cria o pipe
	pid = fork(); // Cria o processo
	if(pid != 0) // Codigo do pai
	{
		printf("Pai vai ler o pipe\n");
		if(read(fd[0], mensagem, 30) < 0) 
			printf("Erro na leitura do pipe\n");
		else
			printf("Pai leu: %s\n", mensagem);
	}
	else // Codigo do filho
	{
		strcpy(mensagem, "HELLO PIPE");
		printf("Filho vai escrever no pipe\n");
		if (write(fd[1], mensagem, 30) < 0)
			printf("Erro na escrita do pipe\n");
		printf("Filho terminou de escrever no pipe\n");
	}
	return 0;
}

File: 2_Programação_sistemas_operacionais/2.4_Pipes_Sinais_Alarmes/Python/Ex2.py
import os

fd = os.pipe()
pid_filho = os.fork()
if pid_filho != 0:
	print("Pai vai ler o pipe")
	mensagem = os.read(fd[0], 30)
	if(len(mensagem)==0):
		print("Erro na leitura do pipe")
	else:
		print("Pai leu: " +  mensagem.decode("utf-8"))
else:
	mensagem = "HELLO PIPE"
	mensagem += "\0"*(30-len(mensagem)) 
	print("Filho vai escrever no pipe")
	if os.write(fd[1], str.encode(mensagem)) < 0:
		print("Erro na escrita do pipe")
	print("Filho terminou de escrever no pipe")